### DISCUSSION TITLE
Detecting Power-On-Reset

### DISCUSSION THREAD

**Discussion by 234d** [troglobit/watchdogd#42] [General]
I might be missing something. But why is `WDIOF_POWERUNDER` used as a flag to detect (from my interpretation) a POR? There are very few watchdog drivers that support this is the kernel, four as bootstatus and two as runtime status. A watchdog that is triggered from a low power situation isn't necessarily the same as a POR. Looking at the watchdog drivers in the kernel they return 0 for a POR. This is also corroborated by the [watchdog test](https://github.com/torvalds/linux/blob/master/tools/testing/selftests/watchdog/watchdog-test.c#L159) in the kernel.

**Comment by troglobit** [comment]
`WDIOF_POWERUNDER` is used to detect "Power bad/power fault", i.e. when loss of power caused the system to reboot. There are other drivers than those in the kernel.

**Comment by 234d** [comment]
My understanding is that `WDIOF_POWERUNDER` is just a fault among the rest. The documentation for the flags in the kernel isn't very exhaustive so not entirely clear how it should be used. And can't find any good documentation on the drivers in kernel that use this feature. Is there any public out-of-tree driver you can point me to that implements this functionality? What I'm thinking is that a watchdog can be set to monitor power in and if it falls below a certain threshold it triggers the watchdog and saves that information for next boot, or can just be read via the ioctl. Which is certainly different than a POR. And getting an "Reset cause" of 0x0000 as [WDIOF_UNKNOWN](https://elixir.bootlin.com/linux/latest/source/include/uapi/linux/watchdog.h#L36) on POR is just wrong I'd say.

**Reply by troglobit** [reply]
It's a capability flag and a reset reason, a good example is `pm8916_wdt.c`.   It is supposed to be used for systems that can detect power failure (loss of power-good signal) when coming back up.  

Not sure I can help you more than that.

**Reply by 234d** [reply]
Thanks for your quick replies and patience!
I'm totally with you and have of course looked at that driver.
What I still don't understand is why it's used in the way it is. Why zero the reset counter on it and not one of the other reset reasons? I would find it more reasonable to reset it on a POR/cold boot, i.e. a `bootstatus` of 0 when no watchdog resets have occurred.
There might be something obvious I'm missing here.
And it seems we have reached an impasse. Thanks for your time and I'll get out of your hair.
